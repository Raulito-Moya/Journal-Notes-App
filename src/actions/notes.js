import Swal from 'sweetalert2';

import { db } from "../firebase/firebase-config";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";




export const startNewNote =() => {
    return async (dispatch, getState) =>{  //getState es una funcione para obtener el sstate

        const uid = getState().auth.uid //obtnego el uid
       // console.log(uid);

       const newNote = {
           title: '',
           body: '',
           date: new Date().getTime() //me va adar el momento exacto en que la persona crea esta nota
       }


    const doc = await db.collection(`${ uid }/journal/notes`).add(newNote);
  
    // console.log(doc);

     dispatch( activeNote( doc.id, newNote) )
     dispatch( addNewNote( doc.id, newNote) )

    }


};


export const activeNote = (id, note) => ({ //para marcar la nota activa
      type: types.notesActive,
      payload: {
          id,
          ...note
      }
 
});

export const addNewNote = ( id, note ) => ({ //agregar la nueva nota al sidebar
     type: types.notesAddNew,
     payload:{
      id,
      ...note
     }

})


export const startLoadingNotes = ( uid ) => {
     return async(dispatch) =>{
 
        const notes = await loadNotes(uid)  //vamos a cargar las notas  de la base de datos
        dispatch(setNotes(notes))
     }
  

}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
})



export const startSaveNote = (note) => { //para guardar en la base de datos
   return async(dispatch,getState) => {
        //console.log(note);
        const {uid} = getState().auth;

       if( !note.url ){
           delete note.url; //borro la propiedad url
       }

        const noteToFirestore = {...note};
        delete noteToFirestore.id;
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore) //se guarda la nota
     
   
        dispatch(refreshNote( note.id, noteToFirestore ));
         Swal.fire('Saved', note.title, 'success');
       }

}

export const refreshNote = ( id, note ) => ({  //para cargar la accion especifica en el menu ya que si se hace con startLoadingNotes se estarian cargando todas
  type: types.notesUpdated,
  payload:{
     id, 
     note: {
         id,
         ...note
     }
   }
})



export const startUpLoading = ( file ) => {
  return async( dispatch, getState ) =>{
     
     const {active:activeNote} = getState().notes;
 
    // console.log(file);
   //  console.log(activeNote);

    Swal.fire({  //va a disparar el cartel de cargando
        title:'Uploading...',
        text:'Please wait...',
        allowOutsideClick: false,
        onBeforeOpen: ()=> {
            Swal.showLoading();
        }
    });

     const fileUrl = await fileUpload(file)
     activeNote.url = fileUrl //le actualizo la url

     console.log(fileUrl);

     dispatch(startSaveNote(activeNote))

     Swal.close();
  }

}


export const startDeleting = (id) => {
     return async(dispatch, getState) =>{

        const uid = getState().auth.uid
        await db.doc(`${uid}/journal/notes/${id}`).delete()
  
      dispatch(deleteNote(id))

     }


}

export const deleteNote =(id) =>({
    type: types.notesDelete,
    payload: id

})

export const noteLogout = () => ({ //borrar las notas de memoria cuando el usuario se desloguee
     type: types.notesLogoutCleaning,
     payload: []
})