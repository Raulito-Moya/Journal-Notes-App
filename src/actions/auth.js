import Swal from 'sweetalert2'
import withReactContent from 'sweetalert2-react-content'

import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { types } from "../types/types";
import { noteLogout } from './notes';
import { finishLoading, startLoading } from './ui';
 
const MySwal = withReactContent(Swal) //para los efectos de aniamacion en el form

export const startLoginEmailPassword = (email,password) => { //aqui voy a simular una accion asincrona
     return (dispatch) => {
        
       dispatch(startLoading())
     return  firebase.auth().signInWithEmailAndPassword(email,password) //esta instruccion es para logear al usuario
              .then( ({user}) => {
                 // console.log(user)
               dispatch( login(user.uid, user.displayName))  
               dispatch(finishLoading())
            })
            .catch(e => {
               console.log(e); 
               dispatch(finishLoading()) //por si hay un error
               MySwal.fire( 'Error', e.message , 'error')
              })

     }

};

export const startRegisterRegisterWithEmailPasswordName = ( email, password, name) => {

   return( dispatch ) => {

      firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({user}) =>{

            await user.updateProfile({ displayName: name }) //update user profile data, se espera que se actualize el objeto para despues hacr el dispatch

            //  console.log(user);
                  dispatch( 
                       login(user.uid, user.displayName) 
                      ) 
             })
             .catch(e =>{
               console.log(e);
               MySwal.fire( 'Error', e.message , 'error')
             })
   }



};

export const startGoogleLogin = () => {  //aqui hago un sign in con google
   return (dispatch) => {
  
     firebase.auth().signInWithPopup(googleAuthProvider)
           .then( ({user}) =>{
             dispatch( login(user.uid, user.displayName) ) 
           })
   }


};

export const login = (uid, displayName) => ({
       type: types.login,
       payload: {
         uid,
         displayName
       }
});


export const startLogout = () => {
     return async( dispatch ) => {

      await firebase.auth().signOut();
      dispatch(logout())
      dispatch(noteLogout()) //
     }

}


export const logout = () => ({
   type: types.logout

})