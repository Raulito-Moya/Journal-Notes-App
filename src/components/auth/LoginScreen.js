import React from 'react';
import {Link} from 'react-router-dom';
import {useDispatch, useSelector} from 'react-redux';
import { login, startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'; //aqui traigo la accion
import {setError, removeError} from '../../actions/ui'
import validator from 'validator';
import { useForm } from '../../hooks/useForm';


export const LoginScreen = () => {

     const dispacth = useDispatch()   //el useDispatch es para hacer el dispatch de acciones
     const {msgError, loading} = useSelector(state=> state.ui)
    
    // console.log(state);


     const [formValues, handleInputChange, reset] = useForm({
       email: 'nando8moya@gmail.com',
       password: '123456'
     });

    const {email,password} = formValues;



    const handleLogin = (e) =>{
      e.preventDefault()

      
         dispacth(startLoginEmailPassword(email,password)) //aqui va a disparar la accion asincrona
         //console.log(email, password);
      
    };


    const handleGoogleLogin = () => {
    
    if(isFormValid()){ 
       dispacth(startGoogleLogin()) 
      }  
      console.log('hey!!');
    };




    const isFormValid = () => {

        if( !validator.isEmail(email)){
         dispacth(setError('email is not valid'))
          console.log('email is not valid')
          return false;
   
      } else if( password.length < 6  ){
         dispacth(setError('wrong password'))
        console.log('Password should be at least 6 charecters and match each other');
        return false
      }
       
      dispacth(removeError())
      return true 
     
    }
   

 return(
     <>
         <h3 className="auth__title">Login</h3>

         <form 
         onSubmit={handleLogin}
        className="animate__animated animate__fadeIn animate__faster"  >

           { 
            msgError && ( 
             <div className="auth__alert-error">
               {msgError}
           </div>)
           
           }  

             <input 
             type="text" 
             placeholder="Email"
             name="email" 
             value={email}
             className="auth__input"
             autoComplete="off"
             onChange={handleInputChange}/>

            <input 
             type="password" 
             placeholder="password"
             name="password" 
             value={password}
              className="auth__input"
              onChange={handleInputChange}/>

         
             <button 
               type="submit"
               className="btn btn-primary btn-block"
               disabled={loading}
               >
                Login
            </button>

            <hr/>
            
           <div className="auth_social-networks">
               <p>Login with social networks</p>

               <div 
                  className="google-btn"
                  onClick={handleGoogleLogin}
                 >
               <div className="google-icon-wrapper">
                  <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
               </div>
                 <p className="btn-text">
                   <b>Sign in with google</b>
                </p>
               </div>
           </div>
   
           <Link 
           to='/auth/register'
           className="link"
           >
             Create New Acount  
           </Link>
            
         </form>
     </>
 )



}