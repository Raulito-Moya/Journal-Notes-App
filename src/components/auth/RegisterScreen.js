import React from 'react';
import {useDispatch, useSelector} from 'react-redux'
import {Link} from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator  from 'validator'; //Aqui implemento validator paa validar los campos
import {setError,removeError} from '../../actions/ui'
import { startRegisterRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen =()=>{

  const dispatch = useDispatch();
  const { msgError } = useSelector( state => state.ui) //el useSelector es para seleccionar el estado actual y usarlo en la vista

  console.log(msgError);

  const [formValues, handleInputChange, reset] = useForm({
     name:'Nando',
     email:'nando8moya@gmail.com',
     password:'123456',
     password2:'123456'
   })

 const {name,email,password,password2} =formValues;

  
 const handleRegister = (e) => {
    e.preventDefault()
   
    if (isFormValid()) {
      dispatch(startRegisterRegisterWithEmailPasswordName(email, password, name));
    }

 }

 const isFormValid = () => {

   if(name.trim().length === 0 ) {
      dispatch(setError('name is required'))
      console.log('name is required')
      return false;

   } else if( !validator.isEmail(email)){
      dispatch(setError('email is not valid'))
       console.log('email is not valid')
       return false;

   } else if( password !== password2 || password.length < 5){
      dispatch(setError('Password should be at least 6 charecters and match each other'))
     console.log('Password should be at least 6 charecters and match each other');
     return false
   }
    
   dispatch(removeError())
   return true 
  
 }







   return(
    <div>
         
      <h3 className="auth__title">Register</h3>

       <form 
        onSubmit={handleRegister}
        className="animate__animated animate__fadeIn animate__faster">

       { 
         msgError && 
        ( 
          <div className="auth__alert-error">
             {msgError}
          </div>
         )
        }

           <input 
           type="text" 
           placeholder="name"
           name="name" 
           value={name}
           onChange={handleInputChange}
           className="auth__input"
           autoComplete="off"/>
     
          <input 
           type="text" 
           placeholder="email"
           name="email" 
           value={email}
           onChange={handleInputChange}
           className="auth__input"
           autoComplete="off"/>
  

      
          <input 
              type="password" 
              placeholder="password"
              name="password" 
              value={password}
              onChange={handleInputChange}
               className="auth__input"/>
          

             <input 
              type="password" 
              placeholder="Confirm Password" 
              name="password2" 
              value={password2}
              onChange={handleInputChange}
              className="auth__input"/>
          
              <button 
                type="submit"
                className="btn btn-primary btn-block mb-5"
                disabled={false}
                >
              Register
             </button>
          
          
          
            <Link 
            to='/auth/login'
            className="link"
            >
              Alredy Registered
            </Link>
             
          </form>
                 </div>
             )
  
  
}