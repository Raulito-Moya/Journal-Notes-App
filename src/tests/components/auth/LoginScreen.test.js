import React from 'react';
import {mount} from 'enzyme';
import {Provider} from 'react-redux'
import { MemoryRouter } from 'react-router-dom' //me va a permitir fingir las rutas
import { LoginScreen } from "../../../components/auth/LoginScreen";
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { startGoogleLogin, startLoginEmailPassword } from '../../../actions/auth';


jest.mock('../../../actions/auth',()=>({
   startGoogleLogin: jest.fn(),
   startLoginEmailPassword: jest.fn()
}))


const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initialstate = {
   auth: {},
   ui: {
      loading: false,
      msgError: null
   }
}

   let store = mockStore(initialstate) 
    store.dispatch = jest.fn() //vamos aremplazar la funcionalidad que tenga el store por una funcion

describe('Pruebas en el LoginScreen', () => {
   
   beforeEach(()=>{
      store = mockStore(initialstate);
      jest.clearAllMocks();  //limpiar los mocks
  })

  const wrapper = mount(  //aqui e proveo el redux con el provider
   <Provider store={store}> 
   <MemoryRouter>
    <LoginScreen/>
   </MemoryRouter>
   </Provider>
  )



   test('Pruebas en el <loginScreen/>',()=>{

   

    expect(wrapper).toMatchSnapshot();

   })

   test('debe de disparar la accion de <startLoginScreen/>',()=>{
    
    
      wrapper.find('.google-btn').prop('onClick')();
 
      expect( startGoogleLogin ).toHaveBeenCalled()
   })


   test('debe de disparar el startLoading con sus respectivos argumentos',()=>{
      

    wrapper.find('form').prop('onSubmit')(
       { preventDefault(){} }
    );
     
    
     expect( startLoginEmailPassword).toHaveBeenLastCalledWith('nando8moya@gmail.com','123456')
   })

});