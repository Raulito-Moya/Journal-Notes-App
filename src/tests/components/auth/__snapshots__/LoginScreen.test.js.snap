// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el LoginScreen Pruebas en el <loginScreen/> 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Login
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="nando8moya@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <hr />
    <div
      className="auth_social-networks"
    >
      <p>
        Login with social networks
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link"
      href="/auth/register"
      onClick={[Function]}
    >
      Create New Acount
    </a>
  </form>,
]
`;
